File: pyforms_web/web/static/pyformsjs/ControlLineChart.js
class ControlLineChart extends ControlBase {

    ////////////////////////////////////////////////////////////////////////////////

    init_control() {
        this.update_it = false;

        var html = "<div id='" + this.place_id() + "' class='field control ControlLineChart' >";
        if (this.properties.label_visible) html += "<label for='" + this.control_id() + "'>" + this.properties.label + "</label>";
        html += "<div id='" + this.control_id() + "' title='" + this.properties.help + "' style='margin-left: 20px' ></div>";
        html += "</div>";
        this.jquery_place().replaceWith(html);

        if (this.properties.value) {
            this.set_value(this.properties.value);
        }

        if (this.properties.required) this.set_required();
    };


    ////////////////////////////////////////////////////////////////////////////////

    set_value(data) {
        if( !this.jquery( ).length ) return;

        if (this.chart) {
            this.chart.destroy();
        }

        var legend = this.properties.legend;

        if (!data || data.length == 0 || data[0].length == 0) {
            data = [[[0, 0]]];
        }

        console.debug(data);
        console.debug(this.properties);

        this.chart = $.jqplot(this.control_id(), data, {
            height: this.properties.height,
            width: this.properties.width,
            seriesColors: [
                '#f2711c', '#fbbd08', '#b5cc18', '#21ba45', '#00b5ad',
                '#2185d0', '#6435c9', '#a333c8', '#e03997', '#a5673f',
                '#767676', '#1b1c1d', '#DB2828'
            ],
            grid: {
                borderColor: 'transparent',
                shadow: false,
                drawBorder: false,
                shadowColor: 'transparent',
                background: 'transparent'
            },
            title: this.label,
            seriesDefaults: {
                showMarker: this.properties.show_marker, 
                showLine: true, 
                lineWidth: 1,
                markerOptions: {size: 4},
                rendererOptions: {
                    smooth: this.properties.smooth
                }
            },
            legend: {
                show: legend.length > 0,
                labels: legend,
                placement: this.properties.legend_placement,
                location: this.properties.legend_location
            },
            axes: {
                xaxis: {
                    renderer: this.properties.x_axis_type==='datetime'?$.jqplot.DateAxisRenderer:$.jqplot.LinearAxisRenderer,
                    tickOptions: {formatString: this.properties.x_axis_format},
                    min: this.properties.xaxis_range? new Date(this.properties.xaxis_range[0]): null,
                    max: this.properties.xaxis_range? new Date(this.properties.xaxis_range[1]): null,
                }
            },
            cursor: {
                style: 'pointer',
                show: true,
                zoom: true,
                tooltipOffset: 10,
                showTooltip: true,
                followMouse: true,
                showTooltipDataPosition: true,
                showVerticalLine: true
            }
        });

        const self = this;
        this.jquery().bind('jqplotDataClick',
            function (ev, seriesIndex, pointIndex, data) {
                self.properties.selected_series = seriesIndex;
                self.properties.selected_data = data;
                self.update_it = true;
                self.basewidget.fire_event(self.name, 'remote_data_selected_event');
            }
        );

        if (this.properties.visible)
            this.show();
        else
            this.hide();
    };

    ////////////////////////////////////////////////////////////////////////////////

    update_server() {
        return this.get_value() != this.properties.value || this.update_it;
    }

    ////////////////////////////////////////////////////////////////////////////////

    deserialize(data) {
        this.properties = $.extend(this.properties, data);
    };

    apply_deserialization(data) {
        super.apply_deserialization(data);
        this.set_value(this.properties.value);
    }

    ////////////////////////////////////////////////////////////////////////////////

    serialize() {
        this.properties.value = this.get_value();
        return this.properties;
    };
}

	

File: pyforms_web/controls/control_linechart.py
import datetime

from pyforms_web.basewidget import custom_json_converter
from pyforms_web.controls.control_base import ControlBase
import simplejson

class ControlLineChart(ControlBase):

    def __init__(self, *args, **kwargs):
        self._legend = []
        self.selected_data = None
        self.selected_serie = None
        self.height = kwargs.get('height', None)
        self.width = kwargs.get('width', None)
        self.legend_location = kwargs.get('legend_location', 'e')
        self.legend_placement = kwargs.get('legend_placement', 'outside')
        self.x_axis_format = kwargs.get('x_axis_format', None)
        self.x_axis_type = kwargs.get('x_axis_type', 'datetime')
        self.smooth = kwargs.get('smooth', True)
        self.show_marker = kwargs.get('show_marker', True)

        self._xaxis_range = kwargs.get('xaxis_range', None)

        super().__init__(*args, **kwargs)

        self.data_selected_event = kwargs.get('data_selected_event', self.data_selected_event)

    def init_form(self): return "new ControlLineChart('{0}', {1})".format( self._name, simplejson.dumps(self.serialize(), default=custom_json_converter) )

    def remote_data_selected_event(self):
        self.data_selected_event(self.selected_serie, self.selected_data)

    def data_selected_event(self, series_index, data):
        pass

    @property
    def xaxis_range(self): return self._xaxis_range

    @xaxis_range.setter
    def xaxis_range(self, value): self._xaxis_range = value

    def serialize(self):
        data = ControlBase.serialize(self)
        
        legend, series = [], []
        value = self._value if isinstance(self._value, dict) else {}

        for title, serie in value.items():
            legend.append(title)
            series.append(serie)

        data.update({ 
            'legend': legend,
            'legend_location': self.legend_location,
            'legend_placement': self.legend_placement,
            'value':  series,
            'height': self.height,
            'width': self.width,
            'x_axis_format': self.x_axis_format,
            'x_axis_type': self.x_axis_type,
            'smooth': self.smooth,
            'show_marker': self.show_marker,
            'xaxis_range': self.xaxis_range,
        })

        return data

    def deserialize(self, properties):
        pass